pkg/cloud/services/provider/provider.go

```diff
@@ -170,7 +170,16 @@ type projects struct {
 	Projects []project `json:"projects"`
 }
 
-func GetProjectID(client *gophercloud.ProviderClient, name string) (string, error) {
+func GetProjectID(client *gophercloud.ProviderClient, clientOpts *clientconfig.ClientOpts) (string, error) {
+	if clientOpts.AuthInfo.ProjectID != "" {
+		return clientOpts.AuthInfo.ProjectID, nil
+	}
+
+	projectName := clientOpts.AuthInfo.ProjectName
+	if projectName == "" {
+		return "", fmt.Errorf("failed to get project id")
+	}
+
```

Comment:
Good call moving this block of code into this function.

Reply:
I think we're going to have to fix this in order to support application credentials, btw, which is another reason to centralise it.

cc @pierreprinetti